android/src/main/java/com/amplitude/android/plugins/AndroidLifecyclePlugin.kt

```diff
@@ -2,35 +2,80 @@ package com.amplitude.android.plugins
 
 import android.app.Activity
 import android.app.Application
+import android.content.pm.PackageInfo
+import android.content.pm.PackageManager
 import android.os.Bundle
 import com.amplitude.android.Configuration
+import com.amplitude.android.utilities.DefaultEventUtils
 import com.amplitude.core.Amplitude
 import com.amplitude.core.platform.Plugin
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicInteger
 
 class AndroidLifecyclePlugin : Application.ActivityLifecycleCallbacks, Plugin {
     override val type: Plugin.Type = Plugin.Type.Utility
     override lateinit var amplitude: Amplitude
+    private lateinit var packageInfo: PackageInfo
+
+    private val hasTrackedApplicationLifecycleEvents = AtomicBoolean(false)
+    private val numberOfActivities = AtomicInteger(1)
+    private val isFirstLaunch = AtomicBoolean(false)
 
     override fun setup(amplitude: Amplitude) {
         super.setup(amplitude)
-        ((amplitude.configuration as Configuration).context as Application).registerActivityLifecycleCallbacks(this)
+
+        val application = (amplitude.configuration as Configuration).context as Application
+        val packageManager: PackageManager = application.packageManager
+        packageInfo = try {
+            packageManager.getPackageInfo(application.packageName, 0)
+        } catch (e: PackageManager.NameNotFoundException) {
+            // This shouldn't happen, but in case it happens, fallback to empty package info.
+            amplitude.logger.error("Cannot find package with application.packageName: " + application.packageName)
+            PackageInfo()
+        }
+        application.registerActivityLifecycleCallbacks(this)
     }
 
     override fun onActivityCreated(activity: Activity, bundle: Bundle?) {
+        val configuration = amplitude.configuration as Configuration
+        if (!hasTrackedApplicationLifecycleEvents.getAndSet(true) && configuration.trackingAppLifecycleEvents) {
+            numberOfActivities.set(0)
+            isFirstLaunch.set(true)
+            DefaultEventUtils(amplitude as com.amplitude.android.Amplitude).trackAppUpdatedInstalledEvent(packageInfo)
+        }
+        if (configuration.trackingDeepLinks) {
+            DefaultEventUtils(amplitude as com.amplitude.android.Amplitude).trackDeepLinkOpenedEvent(activity)
+        }
     }
 
     override fun onActivityStarted(activity: Activity) {
+        val configuration = amplitude.configuration as Configuration
```

Comment:
Maybe add a convenience method for getting the typed configuration and amplitude value? Seems to be done in a lot of places.

Reply:
Updated, thanks!